lua/lean/infoview.lua

```diff
@@ -592,45 +596,64 @@ function Pin:remove_parent_info(info)
 end
 
 --- Update this pin's current position.
-function Pin:set_position_params(params, delay, lean3_opts)
-  local old_params = self.position_params
-  self.position_params = params
-
-  lean3_opts = vim.tbl_extend("keep", lean3_opts or {}, {changed = not vim.deep_equal(params, old_params)})
-
-  self:update_extmark()
-  self:update(false, delay, nil, lean3_opts)
+---@param params UIParams
+function Pin:set_position_params(params)
+  self:update_extmark(params)
 end
 
 --- Update pin extmark based on position, used when resetting pin position.
-function Pin:update_extmark()
-  local params = self.position_params
+---@param params UIParams
+function Pin:update_extmark(params)
   if not params then return end
 
-  local buf = vim.fn.bufnr(vim.uri_to_fname(params.textDocument.uri))
+  local buf = vim.fn.bufnr(params.filename)
 
   if buf ~= -1 then
-    local line = params.position.line
-    local buf_line = vim.api.nvim_buf_get_lines(buf, line, line + 1, false)[1]
-    local col = buf_line and vim.str_byteindex(buf_line, params.position.character) or 0
-    local end_col = buf_line and ((col < #buf_line) and
-      vim.str_byteindex(buf_line, params.position.character + 1) or col) or 0
-
-    self.extmark = vim.api.nvim_buf_set_extmark(buf, extmark_ns,
-      line, col,
-      {
-        id = self.extmark;
-        end_col = end_col;
-        hl_group = self.extmark_hl_group;
-        virt_text = self.extmark_virt_text;
-        virt_text_pos = "right_align";
-      })
-    self.extmark_buf = buf
+    local line = params.row
+    local col = params.col
+
+    self:__update_extmark_style(buf, line, col)
+
+    self:update_position()
   end
 end
 
---- Update pin position based on extmark, used when changing text.
-function Pin:update_position(delay, lean3_opts)
+function Pin:__update_extmark_style(buf, line, col)
+  if not buf and not self.extmark then return end
+
+  -- not a brand new extmark
+  if not buf then
+    buf = self.extmark_buf
+    local extmark_pos = vim.api.nvim_buf_get_extmark_by_id(buf, extmark_ns, self.extmark, {})
+    line = extmark_pos[1]
+    col = extmark_pos[2]
+  end
+
+  local buf_line = vim.api.nvim_buf_get_lines(buf, line, line + 1, false)[1]
+  local end_col = 0
```

Comment:
I'm not sure I understand the logic here.  Is the code finding the closest valid UTF-8 position?  Does this actually happen (that we get an invalid UTF-8 position)?  If so, please explain this in a comment.

Reply:
Added a comment. This gets the UTF-8 index of the next codepoint.